plugins/ptp_operator/metrics/manager.go

```diff
@@ -0,0 +1,266 @@
+	delete(p.Ptp4lConfigInterfaces, key)
+	p.lock.Unlock()
+}
+
+//PublishClockClassEvent ...publish events
+func (p *PTPEventManager) PublishClockClassEvent(clockClass float64, eventResourceName string, eventType ptp.EventType) {
+	source := fmt.Sprintf("/cluster/%s/ptp/%s", p.nodeName, eventResourceName)
+	data := ceevent.Data{
+		Version: "v1",
+		Values: []ceevent.DataValue{{
+			Resource:  string(p.publisherTypes[eventType].Resource),
+			DataType:  ceevent.METRIC,
+			ValueType: ceevent.DECIMAL,
+			Value:     clockClass,
+		},
+		},
+	}
+	p.publish(data, source, eventType)
+}
+
+//PublishEvent ...publish events
+func (p *PTPEventManager) PublishEvent(state ptp.SyncState, ptpOffset int64, eventResourceName string, eventType ptp.EventType) {
+	// create an event
+	if state == "" {
+		return
+	}
+	source := fmt.Sprintf("/cluster/%s/ptp/%s", p.nodeName, eventResourceName)
+	data := ceevent.Data{
+		Version: "v1",
+		Values: []ceevent.DataValue{{
+			Resource:  string(p.publisherTypes[eventType].Resource),
+			DataType:  ceevent.NOTIFICATION,
+			ValueType: ceevent.ENUMERATION,
+			Value:     state,
+		}, {
+			Resource:  string(p.publisherTypes[eventType].Resource),
+			DataType:  ceevent.METRIC,
+			ValueType: ceevent.DECIMAL,
+			Value:     float64(ptpOffset),
+		},
+		},
+	}
+	p.publish(data, source, eventType)
+}
+
+func (p *PTPEventManager) publish(data ceevent.Data, resource string, eventType ptp.EventType) {
+	if pubs, ok := p.publisherTypes[eventType]; ok {
+		e, err := common.CreateEvent(pubs.PubID, string(eventType), resource, data)
+		if err != nil {
+			log.Errorf("failed to create ptp event, %s", err)
+			return
+		}
+		if !p.mock {
+			if err = common.PublishEventViaAPI(p.scConfig, e); err != nil {
+				log.Errorf("failed to publish ptp event %v, %s", e, err)
+				return
+			}
+		}
+	} else {
+		log.Errorf("failed to publish ptp event due to missing publisher for type %s", string(eventType))
+	}
+}
+
+// GenPTPEvent ... generate events form the logs
+func (p *PTPEventManager) GenPTPEvent(ptpProfileName string, stats *stats.Stats, eventResourceName string, ptpOffset int64, clockState ptp.SyncState, eventType ptp.EventType) {
+	defer func() {
+		if r := recover(); r != nil {
+			fmt.Println("Recovered in f", r)
+		}
+	}()
+	if clockState == "" {
+		return
+	}
+
+	lastClockState := stats.LastSyncState()
+	threshold := p.PtpThreshold(ptpProfileName)
+	switch clockState {
+	case ptp.LOCKED:
+		switch lastClockState {
+		case ptp.FREERUN: //last state was already sent for FreeRUN , but if its within then send again with new state
+			if isOffsetInRange(ptpOffset, threshold.MaxOffsetThreshold, threshold.MinOffsetThreshold) { // within range
+				log.Infof(" publishing event for ( profile %s) %s with last state %s and current clock state %s and offset %d for ( Max/Min Threshold %d/%d )",
+					ptpProfileName, eventResourceName, stats.LastSyncState(), clockState, ptpOffset, threshold.MaxOffsetThreshold, threshold.MinOffsetThreshold)
```

Comment:
stats.LastSyncState() is called again here. Is it possible it returns a different value than L208?
Better use `lastClockState` from L208 instead.

Reply:
fixed